spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/data/ParquetWithSparkSchemaVisitor.java

```diff
@@ -182,18 +183,15 @@ private static <T> T visitField(
 
   private static <T> List<T> visitFields(
       StructType struct, GroupType group, ParquetWithSparkSchemaVisitor<T> visitor) {
-    StructField[] sFields = struct.fields();
-    Preconditions.checkArgument(
-        sFields.length == group.getFieldCount(), "Structs do not match: %s and %s", struct, group);
     List<T> results = Lists.newArrayListWithExpectedSize(group.getFieldCount());
-    for (int i = 0; i < sFields.length; i += 1) {
-      Type field = group.getFields().get(i);
-      StructField sField = sFields[i];
-      Preconditions.checkArgument(
-          field.getName().equals(AvroSchemaUtil.makeCompatibleName(sField.name())),
-          "Structs do not match: field %s != %s",
-          field.getName(),
-          sField.name());
+    for (StructField sField : struct.fields()) {
+      final int idx;
+      try {
```

Comment:
we should avoid more expensive try-catch for normal condition of null/unknown type. I was thinking about a if check like this.

```
    for (StructField sField : struct.fields()) {
      // skip null types since they don't exist in Parquet
      if (sField.dataType() != DataTypes.NullType) {
        int index = group.getFieldIndex(AvroSchemaUtil.makeCompatibleName(sField.name()));
        Type field = group.getType(index);
        results.add(visitField(sField, field, visitor));
      }
    }
```

Reply:
Actually `GroupType` has this method  that can combine the above two lines into one line
```
  public Type getType(String fieldName) {
    return getType(getFieldIndex(fieldName));
  }
```